BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java

```diff
@@ -7,88 +7,129 @@
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
 import java.util.TreeMap;
+import java.util.stream.Collectors;
 
 import io.bluedb.api.exceptions.BlueDbException;
 import io.bluedb.api.keys.BlueKey;
 import io.bluedb.disk.Blutils;
 import io.bluedb.disk.serialization.BlueSerializer;
 
 public class Segment <T extends Serializable> {
-	private static String SUFFIX = ".segment";
-
-	private final String pathString;
+//	private final String pathString;
 	private final Path path;
 	private final BlueSerializer serializer;
 	
-	public Segment(Path collectionPath, String segmentId, BlueSerializer serializer) {
-		this.path = Paths.get(collectionPath.toString(), segmentId + SUFFIX);
-		this.pathString = this.path.toString();
-		this.serializer = serializer;
-	}
-
-	public Segment(Path collectionPath, long segmentId, BlueSerializer serializer) {
-		this.path = Paths.get(collectionPath.toString(), segmentId + SUFFIX);
-		this.pathString = this.path.toString();
+	public Segment(Path segmentPath, BlueSerializer serializer) {
+		this.path = segmentPath;
+//		this.pathString = this.path.toString();
 		this.serializer = serializer;
 	}
+//
+//	public Segment(Path collectionPath, String segmentId, BlueSerializer serializer) {
+//		this(Paths.get(collectionPath.toString(), segmentId), serializer);
+//	}
+//
+//	public Segment(Path collectionPath, long segmentId, BlueSerializer serializer) {
+//		this(collectionPath, String.valueOf(segmentId), serializer);
+//	}
 
 	public void put(BlueKey key, T value) throws BlueDbException {
-		TreeMap<BlueKey, BlueEntity<T>> data = load();
+		TreeMap<BlueKey, BlueEntity<T>> data = load(key.getGroupingNumber());
 		BlueEntity<T> entity = new BlueEntity<T>(key, value);
 		data.put(key, entity);
+		String pathString = Paths.get(path.toString(), String.valueOf(key.getGroupingNumber())).toString();
 		Blutils.save(pathString, data, serializer);
 	}
 
 	public void delete(BlueKey key) throws BlueDbException {
-		TreeMap<BlueKey, BlueEntity<T>> data = load();
+		TreeMap<BlueKey, BlueEntity<T>> data = load(key.getGroupingNumber());
 		data.remove(key);
+		String pathString = Paths.get(path.toString(), String.valueOf(key.getGroupingNumber())).toString();
 		Blutils.save(pathString, data, serializer);
 	}
 
 	public BlueEntity<T> read(BlueKey key) throws BlueDbException {
-		Collection<BlueEntity<T>> values = load().values();
+		Collection<BlueEntity<T>> values = load(key.getGroupingNumber()).values();

```

Comment:
The load gives you a map. You should just be able to call get and pass in the key instead of iterating through all of the values.

Though, we talked about getting rid of the TreeMap for starters.